### PR TITLE
Adds functionality to monthly contribution dropdown (Panel, #2)

### PR DESCRIPTION
**Pull request by ryanml**
Fixes: https://github.com/brave/brave-browser/issues/2245
Fixes: https://github.com/brave/brave-browser/issues/2823
UI: https://github.com/brave/brave-ui/pull/434

Preview:
<div style="width: 100%">
<div style="float: left; width: 50%">
Screen Shot 2019-03-20 at 6 52 47 PM
</div>
<div style="float: right; width: 50%">
Screen Shot 2019-03-20 at 6 52 57 PM
</div>
</div>

The brunt of the work for this was adding extension functions to support:

Adding a recurring contribution
Removing a recurring contribution
Retrieving recurring contributions

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [ ] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
   1. Enable Brave Rewards from a clean profile, add tokens
   2. Navigate to a site such as `brave.com`
   3. Bring up the wallet panel.
   4. Ensure the monthly contribution dropdown is not shown
   5. Click "Send Tip" and make a monthly donation (mark "Make this monthly")
   6. Re-open the wallet panel
   7. Change monthly contribution amount to a different amount than the recurring tip
   8. Close and re-open panel
   9. Confirm that the value set in step 7 persists
   10. Confirm that the change is reflected in the donation table
   11. Confirm that setting a monthly amount of 0.0 removes the donation from the table.
 
   The above six steps should be tried in different ways with different values. Ex: Ensuring that the data persists across tabs with the same site open

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx [hunk 1]
```diff
@@ -299,21 +341,25 @@ export class Panel extends React.Component<Props, State> {
...
+                tipsEnabled={true}
+                includeInAuto={!publisher.excluded}
+                attentionScore={(publisher.percentage || 0).toString()}
+                onToggleTips={this.doNothing}
+                donationAmounts={this.generateAmounts()}
+                donationAction={this.showDonateToSiteDetail}
+                onAmountChange={this.onContributionAmountChange.bind(this, publisher.publisher_key)}
+                onIncludeInAuto={this.switchAutoContribute}
+              />
+              <div style={{ height: '48px' }}></div>
```
Comment (NejcZdovc) [inline components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: please add comment about this line

Comment (ryanml) [reply components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: @NejcZdovc done

File: test/BUILD.gn [hunk 2]
```diff
@@ -309,6 +310,7 @@ test("brave_browser_tests") {
     ":brave_browser_tests_deps",
     ":browser_tests_runner",
     "//testing/gmock",
+    "//brave/vendor/bat-native-ledger",
```
[the remark this answers is not in the document]
Comment (ryanml) [reply test/BUILD.gn]: @jasonrsadler nope - removed

File: components/brave_rewards/browser/content_site.cc [hunk 3]
```diff
@@ -26,6 +27,7 @@ namespace brave_rewards {
     url = properties.url;
     provider = properties.provider;
     id = properties.id;
+    weight = properties.weight;
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/content_site.cc]: why do we need to add `weight`?

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 4]
```diff
@@ -1549,6 +1563,10 @@ void RewardsServiceImpl::OnTipsUpdatedData(const ledger::PublisherInfoList list)
   }
 }
 
+void RewardsServiceImpl::AddRecurringPayment(const std::string& publisher_key, double new_amount) {
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: can you just use `SaveRecurringDonation` directly, so that we don't create new function?

Comment (ryanml) [reply components/brave_rewards/browser/rewards_service_impl.cc]: done

File: components/brave_rewards/browser/extension_rewards_service_observer_browsertest.cc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    return info;
+  }
+
+  RewardsService* rewards_service_;
+  bool donations_should_be_empty = false;
+};
+
+IN_PROC_BROWSER_TEST_F(ExtensionRewardsServiceObserverBrowserTest, SaveAndRemoveRecurringDonation) {
+  rewards_service_->AddObserver(this);
```
[the remark this answers is not in the document]
Comment (ryanml) [reply components/brave_rewards/browser/extension_rewards_service_observer_browsertest.cc]: done

**Comment by NejcZdovc** [comment]
currently blocked as we need to define desired behaviour

File: components/brave_rewards/resources/extension/brave_rewards/background/reducers/rewards_panel_reducer.ts [hunk 6]
```diff
@@ -218,6 +218,31 @@ export const rewardsPanelReducer = (state: RewardsExtension.State | undefined, a
...
         break
       }
+    case types.SAVE_RECURRING_DONATION:
+      {
+        let newAmount = payload.newAmount
+        let publisherKey = payload.publisherKey
+        if (publisherKey == null) {
+          break
+        }
+        chrome.braveRewards.saveRecurringDonation(publisherKey, newAmount)
```
Comment (NejcZdovc) [inline components/brave_rewards/resources/extension/brave_rewards/background/reducers/rewards_panel_reducer.ts]: let's check if amount is number and positive as well

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx [hunk 7]
```diff
@@ -88,6 +88,46 @@ export class Panel extends React.Component<Props, State> {
...
+      this.actions.removeRecurringContribution(publisherKey)
+    } else {
+      this.actions.saveRecurringDonation(publisherKey, newValue)
+    }
+  }
+
+  generateAmounts = () => {
+    const { rates } = this.props.rewardsPanelData.walletProperties
+
+    return [0, 1, 5, 10].map((value: number) => {
```
Comment (NejcZdovc) [inline components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: this values needs to come from publisher list and default back to this if publisher didn't set it

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 8]
```diff
@@ -1822,6 +1823,19 @@ void RewardsServiceImpl::OnRecurringDonationsData(const ledger::PublisherInfoLis
...
 
   callback(list, 0);
+
+  // Incoming ledger::PublisherInfoList needs to be converted to brave_rewards::ContentSiteList
+  brave_rewards::ContentSiteList site_list;
+  for (auto& info : list) {
+    site_list.push_back(PublisherInfoToContentSite(info));
+  }
+
+  TriggerOnRecurringDonations(site_list);
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: This method is the response to `GetRecurringDonations` and so should not trigger an observer methods. Observer methods should only be triggered in response to changes, not in response to queries

Comment (ryanml) [reply components/brave_rewards/browser/rewards_service_impl.cc]: Fixed, This is now done via an asynchronous extension method

File: components/brave_rewards/resources/extension/brave_rewards/background/reducers/rewards_panel_reducer.ts (hunk 6 above)
Comment (ryanml) [reply components/brave_rewards/resources/extension/brave_rewards/background/reducers/rewards_panel_reducer.ts]: done

File: browser/extensions/api/brave_rewards_api.cc [hunk 9]
```diff
@@ -270,5 +270,81 @@ void BraveRewardsGetACEnabledFunction::OnGetACEnabled(bool enabled) {
...
+  RewardsService* rewards_service_ = RewardsServiceFactory::GetForProfile(profile);
+
+  if (rewards_service_) {
+    rewards_service_->SaveRecurringDonation(params->publisher_key, params->new_amount);
+  }
+
+  return RespondNow(NoArguments());
+}
+
+BraveRewardsRemoveRecurringDonationFunction::~BraveRewardsRemoveRecurringDonationFunction() {
```
Comment (jasonrsadler) [inline browser/extensions/api/brave_rewards_api.cc]: line break

File: browser/extensions/api/brave_rewards_api.cc [hunk 10]
```diff
@@ -270,5 +270,81 @@ void BraveRewardsGetACEnabledFunction::OnGetACEnabled(bool enabled) {
...
+}
+
+ExtensionFunction::ResponseAction
+  BraveRewardsRemoveRecurringDonationFunction::Run() {
+
+  std::unique_ptr<brave_rewards::RemoveRecurringDonation::Params> params(
+    brave_rewards::RemoveRecurringDonation::Params::Create(*args_));
+
+  Profile* profile = Profile::FromBrowserContext(browser_context());
+  RewardsService* rewards_service_ = RewardsServiceFactory::GetForProfile(profile);
```
Comment (jasonrsadler) [inline browser/extensions/api/brave_rewards_api.cc]: line break

File: browser/extensions/api/brave_rewards_api.cc [hunk 11]
```diff
@@ -270,5 +270,81 @@ void BraveRewardsGetACEnabledFunction::OnGetACEnabled(bool enabled) {
   Respond(OneArgument(std::make_unique<base::Value>(enabled)));
 }
 
+BraveRewardsSaveRecurringDonationFunction::~BraveRewardsSaveRecurringDonationFunction() {
```
Comment (jasonrsadler) [inline browser/extensions/api/brave_rewards_api.cc]: Line break for lint: No lines > 80 chars

File: browser/extensions/api/brave_rewards_api.cc [hunk 12]
```diff
@@ -270,5 +270,81 @@ void BraveRewardsGetACEnabledFunction::OnGetACEnabled(bool enabled) {
...
+}
+
+ExtensionFunction::ResponseAction
+  BraveRewardsSaveRecurringDonationFunction::Run() {
+
+  std::unique_ptr<brave_rewards::SaveRecurringDonation::Params> params(
+    brave_rewards::SaveRecurringDonation::Params::Create(*args_));
+
+  Profile* profile = Profile::FromBrowserContext(browser_context());
+  RewardsService* rewards_service_ = RewardsServiceFactory::GetForProfile(profile);
```
Comment (jasonrsadler) [inline browser/extensions/api/brave_rewards_api.cc]: line break

File: browser/extensions/api/brave_rewards_api.cc [hunk 13]
```diff
@@ -270,5 +270,81 @@ void BraveRewardsGetACEnabledFunction::OnGetACEnabled(bool enabled) {
...
+
+  if (rewards_service_) {
+    rewards_service_->SaveRecurringDonation(params->publisher_key, params->new_amount);
```
Comment (jasonrsadler) [inline browser/extensions/api/brave_rewards_api.cc]: line break

File: browser/extensions/api/brave_rewards_api.h [hunk 14]
```diff
@@ -155,6 +156,40 @@ class BraveRewardsGetACEnabledFunction : public UIThreadExtensionFunction {
   void OnGetACEnabled(bool enabled);
 };
 
+class BraveRewardsSaveRecurringDonationFunction : public UIThreadExtensionFunction {
```
Comment (jasonrsadler) [inline browser/extensions/api/brave_rewards_api.h]: line break

File: browser/extensions/api/brave_rewards_api.h [hunk 15]
```diff
@@ -155,6 +156,40 @@ class BraveRewardsGetACEnabledFunction : public UIThreadExtensionFunction {
...
+ public:
+  DECLARE_EXTENSION_FUNCTION("braveRewards.saveRecurringDonation", UNKNOWN)
+
+ protected:
+  ~BraveRewardsSaveRecurringDonationFunction() override;
+
+  ResponseAction Run() override;
+};
+
+class BraveRewardsRemoveRecurringDonationFunction : public UIThreadExtensionFunction {
```
Comment (jasonrsadler) [inline browser/extensions/api/brave_rewards_api.h]: line break

File: browser/extensions/api/brave_rewards_api.h [hunk 16]
```diff
@@ -155,6 +156,40 @@ class BraveRewardsGetACEnabledFunction : public UIThreadExtensionFunction {
...
+ public:
+  DECLARE_EXTENSION_FUNCTION("braveRewards.removeRecurringDonation", UNKNOWN)
+
+ protected:
+  ~BraveRewardsRemoveRecurringDonationFunction() override;
+
+  ResponseAction Run() override;
+};
+
+class BraveRewardsGetRecurringDonationsFunction : public UIThreadExtensionFunction {
```
Comment (jasonrsadler) [inline browser/extensions/api/brave_rewards_api.h]: line break

File: components/brave_rewards/browser/rewards_service.h [hunk 17]
```diff
@@ -188,6 +190,10 @@ class RewardsService : public KeyedService {
 
   static void RegisterProfilePrefs(PrefRegistrySimple* registry);
 
+  virtual void SaveRecurringDonation(const std::string& publisher_key, const int amount) = 0;
```
Comment (jasonrsadler) [inline components/brave_rewards/browser/rewards_service.h]: line break

File: components/brave_rewards/browser/rewards_service.h [hunk 18]
```diff
@@ -71,6 +71,8 @@ using GetRewardsMainEnabledCallback = base::Callback<void(bool)>;
 using ConfirmationsHistoryCallback = base::Callback<void(int, double)>;
 using GetRewardsInternalsInfoCallback = base::OnceCallback<void(
     std::unique_ptr<brave_rewards::RewardsInternalsInfo>)>;
+using GetRecurringDonationsListCallback =
+    base::Callback<void(std::unique_ptr<brave_rewards::ContentSiteList>)>;
```
Comment (jasonrsadler) [inline components/brave_rewards/browser/rewards_service.h]: Needs to be `OnceCallback`

File: components/brave_rewards/browser/rewards_service_impl.h [hunk 19]
```diff
@@ -368,8 +372,8 @@ class RewardsServiceImpl : public RewardsService,
                             const uint32_t date,
                             const std::string& publisher_key,
                             const ledger::REWARDS_CATEGORY category) override;
-  void GetRecurringDonations(
-      ledger::PublisherInfoListCallback callback) override;
+  void GetRecurringDonations(ledger::PublisherInfoListCallback callback) override;
```
Comment (jasonrsadler) [inline components/brave_rewards/browser/rewards_service_impl.h]: line break

File: components/brave_rewards/browser/rewards_service_impl.h [hunk 20]
```diff
@@ -368,8 +372,8 @@ class RewardsServiceImpl : public RewardsService,
...
+  void GetRecurringDonationsList(GetRecurringDonationsListCallback callback) override;
```
Comment (jasonrsadler) [inline components/brave_rewards/browser/rewards_service_impl.h]: line break

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx [hunk 21]
```diff
@@ -331,6 +331,46 @@ export class Panel extends React.Component<Props, State> {
     }
   }
 
+  onContributionAmountChange = (publisherKey: string, newAmount: string) => {
+    const newValue = parseInt(newAmount, 10)
```
Comment (jasonrsadler) [inline components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: For Typescript, I think use:
`const newValue = number(newAmount)`
but I don't know if there's a convention. wdyt?

**Comment by jasonrsadler** [comment]
I feel like I've asked before so forgive me if I have but are we handling clipping in a different PR?
Screen Shot 2019-03-19 at 8 45 36 PM

**Comment by ryanml** [comment]
@jasonrsadler yeah this is something that will be handled separately, I need to re-bump discussion on it. Simplest approach would be to just style an actual `<select>` component, then the options box will auto-float and not clip in to the edges. Else we'll have to use `DOMIntersectionObserver` or something to check when it's clipping and re-position it.

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx (hunk 21 above)
Comment (ryanml) [reply components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: We generally have been sticking to `parseInt` throught the brave rewards client side so I think we should stick with it. If there's some other benefit of `number` we are missing out on we can discuss switching over 👍

Comment (jasonrsadler) [reply components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: If that's what's been used then I would stick with it.

**Comment by ryanml** [comment]
@jasonrsadler line length issues resolved

File: components/brave_rewards/browser/rewards_service.h (hunk 18 above)
Comment (ryanml) [reply components/brave_rewards/browser/rewards_service.h]: Done, now using `OnceCallback`

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx [hunk 22]
```diff
@@ -331,6 +331,46 @@ export class Panel extends React.Component<Props, State> {
...
+      this.actions.removeRecurringContribution(publisherKey)
+    } else {
+      this.actions.saveRecurringDonation(publisherKey, newValue)
+    }
+  }
+
+  generateAmounts = () => {
+    const { rates } = this.props.rewardsPanelData.walletProperties
+
+    return [0, 1, 5, 10].map((value: number) => {
```
Comment (NejcZdovc) [inline components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: we need to get amounts from publisher list and if it's empty default back

**Comment by ryanml** [comment]
Referring to the comment about select above, as a part of this PR I am going to implement a styled native <select> component so that we do not deliver this with a clipping issue. cc: @jasonrsadler

File: components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx (hunk 22 above)
Comment (ryanml) [reply components/brave_rewards/resources/extension/brave_rewards/components/panel.tsx]: @NejcZdovc these are amounts are now retrieved from publisher donation info, if no custom values were set, it defaults to this list.

File: browser/extensions/api/brave_rewards_api.cc [hunk 23]
```diff
@@ -270,5 +270,107 @@ void BraveRewardsGetACEnabledFunction::OnGetACEnabled(bool enabled) {
...
+
+  Profile* profile = Profile::FromBrowserContext(browser_context());
+  RewardsService* rewards_service =
+    RewardsServiceFactory::GetForProfile(profile);
+
+  if (!rewards_service) {
+    return RespondNow(Error("Rewards service is not initialized"));
+  }
+
+  rewards_service->GetPublisherBanner(params->publisher_key);
```
Comment (NejcZdovc) [inline browser/extensions/api/brave_rewards_api.cc]: I know this was not introduced in this PR, but would be awesome if you could add callback to `GetPublisherBanner` and we can remove observer for it. You would need to change `brave_donate_ui` `GetPublisherBanner` as well.

Comment (ryanml) [reply browser/extensions/api/brave_rewards_api.cc]: @NejcZdovc tbh as I was working this out (and it was a bit tricky) I stopped myself to say I think we just need to let this PR move on. The methods we are using to pass information from `rewards_service_impl` -> `ledger_impl` is changing constantly and is something we should address in separate tickets. Recommend we get this PR wrapped up.

**Comment by ryanml** [comment]
Closed in favor of: https://github.com/brave/brave-core/pull/2069

**Pull request closed without merging**